src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs

```diff
@@ -105,29 +108,37 @@ public AuthorizeFilter(string policy)
 
         bool IFilterFactory.IsReusable => true;
 
-        /// <inheritdoc />
-        public virtual async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+        private async Task<AuthorizationPolicy> GetEffectivePolicyAsync(AuthorizationFilterContext context)
         {
-            if (context == null)
-            {
-                throw new ArgumentNullException(nameof(context));
+            if (_effectivePolicy != null) {
+                return _effectivePolicy;
             }
 
             var effectivePolicy = Policy;
-                            
-            var mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
-            if (mvcOptions.CombineAuthorizeFilters)
+
+            if (_mvcOptions == null) {
+                _mvcOptions = context.HttpContext.RequestServices.GetRequiredService<IOptions<MvcOptions>>().Value;
+            }
+
+            if (_mvcOptions.CombineAuthorizeFilters)
             {
-                // Combine all authorize filters into single effective policy that's only run on the first filter
-                if (context.Filters.First(f => f is AuthorizeFilter) != this)
+                // Combine all authorize filters into single effective policy that's only run on the closest filter
+                AuthorizationPolicyBuilder builder = null;
+                for (var i = 0; i < context.Filters.Count; i++)
                 {
-                    return;
+                    if (ReferenceEquals(this, context.Filters[i]))
+                    {
+                        continue;
+                    }
+                    
+                    if (context.Filters[i] is AuthorizeFilter authorizeFilter)
+                    {
+                        builder = builder ?? new AuthorizationPolicyBuilder(effectivePolicy);
+                        builder.Combine(authorizeFilter.Policy);
+                    }
```

Comment:
Why does all this policy computation happen inside the filter instead of in the https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/AuthorizationApplicationModelProvider.cs

Is there any reason I'm missing here for this not to be the case? /cc: @rynowak 

The only thing I can think of is someone pluging in an authorize filter dynamically through a filter provider but that sounds very far fetched and something we would not recommend our customers.

IMO resolving the effective policy should happen at the application model level and the filter should only make a decision on whether it should execute or not.

Reply:
Ignore this comment. Let's leave this here for now and get the ball rolling, we'll file an issue for this and move it later.